### ISSUE TITLE
PdfJS Integration > Accented chars in IE

### ISSUE THREAD

**Issue by cdujeu** [pydio/pydio-core#854]
Hi @garnetius 

// Opening an issue to be independant of the PR.

I had a quick test run, I did not check the code yet. Seems to be working good, panel resize is working (ok), tested filenames with utf8 or url characters (ok). Still, I detected some issues, see below.

*Blocking*
 * Cannot open two PDF at the same time: probably due to the way you use the global DEFAULT_URL variable. We should be able to pass a URL directly to the webViewerInitialized initialized function.
 * Shared file: preview is not working, there's a trick with the minisite_session parameter to add (see other editors like video)

*Major*
 * We loose the thumbnail preview capability: we would have to mix something with imagick for that...
 * How do we handle non-supporting browsers (IE8) with a fallback to imagick

*Minor*
 * i18n issue: the editor label is wrong for me in french (i18n cache cleared already)
 * Why rewriting a specific read_pdf_data action ? The standard "get_content" action (that is used for downloading a file) should be enough. Working for me with some basic tests. The PHP class would be unnecessary (although we'll probably want to use it to handle browser agent and plugin activity).

I actually ended up merging the PR, except that I re-enabled pdf extension inside imagemagick for the moment. So maybe you can make sure to restart a new PR from the develop branch that now contains your plugin. I can have a look at the "Shared file" case, I let your the "open multiple" issue, right ? 

Charles

**Comment by garnetius** [comment]
Hi @cdujeu,

**Blocking**

  * *Cannot open two PDFs at the same time.* — This might be also due to markup. Yes, opening multiple viewers is a must, I will check it out.

  * *Shared file: preview is not working.* — I will look into this one, should be easy quick fix.

As for the rest, these aren't my area. I basically mess around only with the plugin and whatever API it provides me available with. However, adding additional application logic to split viewers (editors) from previewers (make ImageMagick a previewer, but not viewer), support for browser fallback and so on falls outside of scope I can control. So it's up to you to implement them: PDF.JS is not the only plugin which might need them.

The French label issue is probably due to the fact that only English strings are provided: I've only edited `en.php`.

**Comment by cdujeu** [comment]
About the Shared File, don't spend too much time, I can take it :-)

**Comment by garnetius** [comment]
And I’ve done it. Both multiple viewers and shared preview work just fine now. The shared file issue was not due to `minisite_session`: it failed because of slightly incorrectly constructed URL.

While looking at `viewer.js` for available options to update it for multiple viewers, I’ve discovered a whole bunch of (undocumented) API options it supports through the URL of the file. I call them undocumented because the only way to figure them out is to open `viewer.js` and read it thoroughly until you stumble upon the relevant parts — there’s not a hint about them on PDF.JS website.

As it turns out, the integration can be done much better by simply embedding HTML viewer in an `<iframe>` by dynamically adjusting its `src` attribute for full path to viewer’s HTML file with path to PDF as a parameter in a query string. There are other options it accepts through query string, but I didn’t bother with any of them — maybe you will find some of them interesting.

The new version of the plugin requires much less adjustments to original PDF.JS files (I only tweaked `viewer.html` very slightly too hook it up with additional Pydio CSS) and on top of that has fully working print preview.

From what I can see it works absolutely perfect now.

I will open the pull request shortly.

*Why rewriting a specific read_pdf_data action? The standard `get_content` action (that is used for downloading a file) should be enough.* — I wasn’t sure about that. The PDF.JS might expect correct headers and I didn’t know if `get_content` was the right thing to go with. I assumed the (most) correct way to do the loading is to make a separate PHP class which would stream the file with correct headers while looking at the sources of video plugin. I leave the final decision to you on whether to keep PHP PDF class or not.

**Comment by cdujeu** [comment]
@garnetius 
Thank you so much for your work, it's been a while that we wanted to do that....
Everything is fixed now, I just made sure to change the order of Imagick & PdfViewer so that the latter one is the first for opening (but does not provide thumbnails, so Imagick is). I fixed various i18n stuff. 
And as the License allows us to do so, I add the whole library to the package. Should be done at build time, but currently we don't handle that, so let's put it directly in the code.
-c
